frontend/src/components/pages/roles/RoleUpdatePage.tsx

```diff
@@ -0,0 +1,214 @@
+    toDelete,
+  };
+}
+
+const RoleUpdatePage = () => {
+  const toast = useToast();
+  const nav = useNavigate();
+  const queryClient = useQueryClient();
+  const { roleName = '' } = useParams<{ roleName: string }>();
+
+  useEffect(() => {
+    uiState.pageBreadcrumbs = [
+      { title: 'Security', linkTo: '/security' },
+      { title: 'Roles', linkTo: '/security/roles' },
+      { title: roleName, linkTo: `/security/roles/${roleName}/details` },
+      { title: 'Update Role', linkTo: '', heading: '' },
+    ];
+  }, [roleName]);
+
+  // Fetch existing ACL data for the role
+  const { data, isLoading } = useQuery(
+    listACLs,
+    {
+      filter: {
+        principal: `RedpandaRole:${roleName}`,
+      },
+    } as ListACLsRequest,
+    {
+      enabled: !!roleName,
+      select: getAclFromAclListResponse,
+    },
+  );
+
+  const { mutateAsync: createACLMutation } = useMutation(createACL);
+  const { mutateAsync: deleteACLMutation } = useMutation(deleteACLs);
+
+  const updateRoleAclMutation =
+    (actualRules: Rule[], sharedConfig: SharedConfig) => async (_: string, _2: string, rules: Rule[]) => {
+      const currentRules: ACLWithId[] = convertRulesToCreateACLRequests(
+        actualRules,
+        sharedConfig.principal,
+        sharedConfig.host,
+      ).map((r) => ({
+        ...r,
+        id: getIdFromCreateACLRequest(r),
+      }));
+      const newRules: ACLWithId[] = convertRulesToCreateACLRequests(
+        rules,
+        sharedConfig.principal,
+        sharedConfig.host,
+      ).map((r) => ({
+        ...r,
+        id: getIdFromCreateACLRequest(r),
+      }));
+
+      const { toCreate, toDelete } = calculateACLDifference(currentRules, newRules);
+
+      console.log('Rules to create:', toCreate);
+      console.log('Rules to delete:', toDelete);
+
+      const createResults = toCreate.map((r) => createACLMutation(r));
+      const deleteResults = toDelete.map((r) => {
+        const d: DeleteACLsRequest = {
+          filter: {
+            principal: r.principal,
+            resourceType: r.resourceType,
+            resourceName: r.resourceName,
+            host: r.host,
+            operation: r.operation,
+            permissionType: r.permissionType,
+            resourcePatternType: r.resourcePatternType,
+            $typeName: 'redpanda.api.dataplane.v1.DeleteACLsRequest.Filter',
+          },
+          $typeName: 'redpanda.api.dataplane.v1.DeleteACLsRequest',
+        };
+        return deleteACLMutation(d);
+      });
+
+      await Promise.all([...createResults, ...deleteResults]);
+
+      // TODO: handle partial failures
+      toast({
+        status: 'success',
+        description: 'Role ACLs updated successfully',
+      });
+
+      // Invalidate the listACLs query cache to force fresh data on next request
+      await queryClient.invalidateQueries({
+        queryKey: createConnectQueryKey({
+          schema: ACLService,
+          cardinality: 'finite',
+        }),
+      });
```

Comment:
This should be part of a custom hook.

Reply:
I'll move all update operation in that case